File: src/services/cmds.ts
import { invoke } from '@tauri-apps/api/core';

/**
 * @description: Get app settings data
 * @return {Promise<Nvmd.Setting>} Promise-Nvmd.Setting
 */
export async function getSettings() {
	return invoke<Nvmd.Setting>('read_settings');
}

/**
 * @description: Update app settings data
 * @param {Nvmd.Setting} settings app settings data
 * @return {Promise<void>} Promise-void
 */
export async function updateSettings(settings: Nvmd.Setting) {
	return invoke<void>('update_settings', { settings });
}

/**
 * @description: Get the current global node version
 * @param {boolean} fetch whether to read the latest value from the file
 * @return {Promise<string>} node version
 */
export function vCurrent(fetch: boolean = false) {
	return invoke<string>('current', { fetch });
}

/**
 * @description: Set the globally effective node version
 * @param {string} version node version number
 * @return {Promise<void>}	Promise-void
 */
export function vSetCurrent(version: string) {
	return invoke<void>('set_current', { version });
}

/**
 * @description: Get a list of all officially released versions of node
 * @param {boolean} fetch Whether to pull new data from remote services
 * @return {Promise<Nvmd.Versions>}	List of all officially released versions of node
 */
export function versionList(fetch: boolean = false) {
	return invoke<Nvmd.Versions>('version_list', { fetch });
}

/**
 * @description: Get a list of installed nodes
 * @param {boolean} fetch Whether to read the latest data
 * @return {Promise<Array<string>>} An array containing the node version number
 */
export function installedList(fetch: boolean = false) {
	return invoke<Array<string>>('installed_list', { fetch });
}

/**
 * @description: Download Node
 * @param {string} version node version
 * @param {string} arch	node architecture
 * @return {Promise<string>}	The file path where the downloaded node is saved
 */
export function installNode(version: string, arch?: string) {
	return invoke<string>('install_node', { version, arch });
}

/**
 * @description: Cancel the task of downloading node
 * @return {Promise<void>} Promise-void
 */
export function installNodeCancel() {
	return invoke<void>('install_node_cancel');
}

/**
 * @description	uninstall node
 * @param {string} version version number
 * @param {boolean} current whether the version to be uninstalled is the currently used version
 * @returns {Promise<void>} Promise<void>
 */
export function uninstallNode(version: string, current: boolean = false) {
	return invoke<void>('uninstall_node', { version, current });
}

/**
 * @description: Get project list from locale file
 * @param {boolean} fetch whether to read the latest value from the file
 * @return {Promise<Nvmd.Project[]>}
 */
export function projectList(fetch: boolean = false) {
	return invoke<Nvmd.Project[]>('project_list', { fetch });
}

/**
 * @description: Get group list from locale file
 * @param {boolean} fetch whether to read the latest value from the file
 * @return {Promise<Nvmd.Group[]>}
 */
export function groupList(fetch: boolean = false) {
	return invoke<Nvmd.Group[]>('group_list', { fetch });
}

File: src-tauri/src/cmds.rs
use crate::{
    config::{Config, Group, ISettings, NVersion, Project},
    core::{group, node, project},
    ret_err, wrap_err,
};

use anyhow::Result;

type CmdResult<T = ()> = Result<T, String>;

/// get current version
#[tauri::command]
pub fn current(fetch: Option<bool>) -> CmdResult<Option<String>> {
    wrap_err!(node::get_current(fetch))
}

/// set current version
#[tauri::command]
pub async fn set_current(version: Option<String>) -> CmdResult<()> {
    wrap_err!(node::set_current(version).await)
}

/// fetch node version list
#[tauri::command]
pub async fn version_list(fetch: Option<bool>) -> CmdResult<Option<Vec<NVersion>>> {
    wrap_err!(node::get_version_list(fetch).await)
}

/// read node installed version list
#[tauri::command]
pub async fn installed_list(fetch: Option<bool>) -> CmdResult<Option<Vec<String>>> {
    wrap_err!(node::get_installed_list(fetch).await)
}

/// read settings
#[tauri::command]
pub async fn read_settings() -> CmdResult<ISettings> {
    Ok(Config::settings().data().clone())
}

/// update settings
#[tauri::command]
pub async fn update_settings(settings: ISettings) -> CmdResult<()> {
    Config::settings().apply();
    wrap_err!({ Config::settings().data().patch_settings(settings) })?;

    Ok(())
}

/// install node
#[tauri::command]
pub async fn install_node(
    window: tauri::Window,
    version: Option<String>,
    arch: Option<String>,
) -> CmdResult<String> {
    wrap_err!(node::install_node(window, version, arch).await)
}

/// install node
#[tauri::command]
pub async fn install_node_cancel() -> CmdResult<()> {
    wrap_err!(node::install_node_cancel().await)
}

/// uninstall node
#[tauri::command]
pub async fn uninstall_node(version: Option<String>, current: Option<bool>) -> CmdResult<()> {
    if version.is_none() {
        ret_err!("version should not be null");
    }

    let version = version.unwrap();
    wrap_err!(node::uninstall_node(version, current).await)
}

/// get project list
#[tauri::command]
pub async fn project_list(fetch: Option<bool>) -> CmdResult<Option<Vec<Project>>> {
    wrap_err!(project::project_list(fetch).await)
}

/// add projects
#[tauri::command]
pub async fn add_projects(app_handle: tauri::AppHandle) -> CmdResult<Option<Vec<project::PInfo>>> {
    wrap_err!(project::add_projects(app_handle).await)
}

/// get group list
#[tauri::command]
pub async fn group_list(fetch: Option<bool>) -> CmdResult<Option<Vec<Group>>> {
    wrap_err!(group::group_list(fetch).await)
}

/// exit app
#[tauri::command]
pub fn exit_app(app_handle: tauri::AppHandle) {
    app_handle.exit(0);
    std::process::exit(0);
}
